### ISSUE TITLE
ACME: implement SSO challenge (draft-biggs-acme-sso-01)

### ISSUE THREAD

**Issue by paul-snively** [smallstep/certificates#1011]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

In my opinion, step-ca has an opportunity to significantly advance the goal of Zero-Trust Authentication by extending its already-excellent support for OIDC provisioning to support for [draft-biggs-acme-sso-01](https://datatracker.ietf.org/doc/html/draft-biggs-acme-sso-01), thereby enabling supported ACME clients to provide simple and seamless OIDC-to-X.509 authentication.

## Why is this needed?

ACME and step-ca have done wonders in reducing the historic friction of issuing and deploying X.509 certificates, specifically encouraging wider adoption of mTLS authentication. In parallel, the [https://spiffe.io/](SPIFFE) project has similarly begun to provide wide infrastructure support for [Zero-Trust Architecture](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-207.pdf), a goal also near and dear to SmallStep's heart. Like much X.509-based infrastructure, however, SPIFFE seems to focus primarily on node-to-node and workload-to-workload authentication, essentially ignoring end-user authentication. step-ca with draft-biggs-acme-sso-01 support could potentially help close this gap by providing the flow for end-user-facing systems to participate in a standard OIDC authentication flow exactly as they do today, but with the end result being the system's acquisition of an X.509 leaf certificate. Presumably for the use-cases I have mind, additional steps that are out of scope for this enhancement suggestion would then include:

1. Ensuring the issued X.509 certificate is an [X.509-SVID](https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md) (this is probably doable with X.509 templates).
2. Registering the X.509-SVID with an appropriate SPIFFE server, such as [SPIRE](https://spiffe.io/docs/latest/spire-about/), to make it visible via the [Workload API](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Workload_API.md).

The particular use-case I have in mind includes mapping roles all the way from the OIDC identity token to roles defined in PostgreSQL, and using SPIFFE/SPIRE to provide X.509 certificates to PostgreSQL for authentication as in [this spiffe-helper example](https://github.com/spiffe/spiffe-helper/tree/master/examples/postgresql), resulting in an RBAC solution for data in PostgreSQL all the way from OIDC login to connecting to PostgreSQL at all, and from there to what data that role in PostgreSQL has access to.

**Comment by maraino** [comment]
@paul-snively right now, you can configure an OIDC provisioner on step-ca and get certificates for your persona with your email on them. 

That draft looks to me as an attempt of standarize a similar flow over the ACME protocol. The support for ACME clients of that protocol might take some time.

**Comment by dopey** [comment]
Hey @paul-snively :wave: Thanks for opening this issue!

We've actually thrown this idea around internally before. Eventually, we'd love for all of our provisioners to be supported through ACME. However, as @maraino mentioned, we don't see much value in integrating this flow before more of the popular clients support it. If we're the only ones that support it in the short term, then it's not much different than just using an OIDC provisioner.

I've added this to our triage so we can discuss and prioritize internally. Will let you know here if there are any updates.

Cheers :beers:
